src/hangar/cli/io/manage_plugins.py

```diff
@@ -0,0 +1,353 @@
+import os.path
+from glob import glob
+
+
+# The plugin store will save a list of *loaded* io functions for each io type
+# (e.g. 'imread', 'imsave', etc.). Plugins are loaded as requested.
+plugin_store = None
+# Dictionary mapping plugin names to a list of functions they provide.
+plugin_provides = {}
+# The module names for the plugins in `hangar.plugins._plugins`.
+plugin_module_name = {}
+# Meta-data about plugins provided by *.ini files.
+plugin_meta_data = {}
+# For each plugin type, default to the first available plugin as defined by
+# the following preferences.
+preferred_plugins = {
+    # Default plugins for all types (overridden by specific types below).
+    'all': ['pil', 'matplotlib'],
+    'imshow': ['matplotlib'],
+}
+
+
+def _clear_plugins():
+    """Clear the plugin state to the default, i.e., where no plugins are loaded
+
+    """
+    global plugin_store
+    plugin_store = {'imread': [],
+                    'imsave': [],
+                    'imshow': [],
+                    '_app_show': []}
+
+
+_clear_plugins()
+
+
+def _load_preferred_plugins():
+    # Load preferred plugin for each io function.
+    io_types = ['imsave', 'imshow', 'imread']
+    for p_type in io_types:
+        _set_plugin(p_type, preferred_plugins['all'])
+
+    plugin_types = (p for p in preferred_plugins.keys() if p != 'all')
+    for p_type in plugin_types:
+        _set_plugin(p_type, preferred_plugins[p_type])
+
+
+def _set_plugin(plugin_type, plugin_list):
+    for plugin in plugin_list:
+        if plugin not in available_plugins:
+            continue
+        try:
+            use_plugin(plugin, kind=plugin_type)
+            break
+        except (ImportError, RuntimeError, OSError):
+            pass
+
+
+def reset_plugins():
+    _clear_plugins()
+    _load_preferred_plugins()
+
+
+def _parse_config_file(filename):
+    """Return plugin name and meta-data dict from plugin config file."""
+    parser = ConfigParser()
+    parser.read(filename)
+    name = parser.sections()[0]
+
+    meta_data = {}
+    for opt in parser.options(name):
+        meta_data[opt] = parser.get(name, opt)
+
+    return name, meta_data
+
+
+def _scan_plugins():
+    """Scan the plugins directory for .ini files and parse them
+    to gather plugin meta-data.
+
+    """
+    pd = os.path.dirname(__file__)
+    config_files = glob(os.path.join(pd, '_plugins', '*.ini'))
+
+    for filename in config_files:
+        name, meta_data = _parse_config_file(filename)
+        plugin_meta_data[name] = meta_data
+
+        provides = [s.strip() for s in meta_data['provides'].split(',')]
+        valid_provides = [p for p in provides if p in plugin_store]
+
+        for p in provides:
+            if p not in plugin_store:
+                print(f"Plugin: {name} wants to provide non-existent: {p}. Ignoring.")
+
+        plugin_provides[name] = valid_provides
+        plugin_module_name[name] = os.path.basename(filename)[:-4]
+
+
+_scan_plugins()
```

Comment:
would be great (for readability) if all the function calls are placed together at one place. Or may be we could do it when we clean it up as a proper plugin system

Reply:
Let's leave it. This is going to need some work in the future anyways.